merlion/models/forecast/deep_base.py

```diff
@@ -0,0 +1,289 @@
+        )
+
+        train_steps = len(training_dataset)
+        logger.info(f"Training steps each epoch: {train_steps}")
+
+        bar = ProgressBar(total=config.num_epochs)
+
+        if config.early_stop_patience:
+            early_stopping = EarlyStopping(patience=config.early_stop_patience)
+
+        # start training
+        for epoch in range(config.num_epochs):
+            train_loss = []
+
+            self.deep_model.train()
+            epoch_time = time.time()
+
+            for i, batch in enumerate(training_dataset):
+                self.optimizer.zero_grad()
+
+                loss, _, _ = self._get_batch_model_loss_and_outputs(batch)
+                train_loss.append(loss.item())
+
+                loss.backward()
+                if config.clip_gradient is not None:
+                    torch.nn.utils.clip_grad_norm(self.model.parameters(), config.clip_gradient)
+
+                self.optimizer.step()
+
+            train_loss = np.average(train_loss)
+            _, val_loss = self._get_np_loss_and_prediction(validation_dataset)
+
+            if bar is not None:
+                bar.print(
+                    epoch + 1, prefix="", suffix=f"Train Loss: {train_loss: .4f}, Validation Loss: {val_loss: .4f}"
+                )
+
+            if config.early_stop_patience:
+                early_stopping(val_loss, self.deep_model)
+                if early_stopping.early_stop:
+                    logger.info(f"Early stopping with {config.early_stop_patience} patience")
+                    break
+
+        if config.early_stop_patience:
+            early_stopping.load_best_model(self.deep_model)
+            logger.info(f"Load the best model with validation loss: {early_stopping.val_loss_min: .4f}")
+
+        logger.info("End of the training loop")
+
+        pred, _ = self._get_np_loss_and_prediction(training_dataset)
+
+        # since the model predicts multiple steps, we concatenate all the first steps together
+        return pd.DataFrame(pred[:, 0], columns=total_data.columns), None
+
+    def _get_batch_model_loss_and_outputs(self, batch):
+        """
+        For loss calculation and output prediction
+
+        :param batch: a batch contains `(past, past_timestamp, future, future_timestamp)` used for calculating loss and outputs
+
+        :return: calculated loss, deep model outputs and targeted ground truth future
+        """
+        config = self.config
+        device = self.deep_model.device
+
+        past, past_timestamp, future, future_timestamp = batch
+
+        past = torch.tensor(past, dtype=torch.float, device=device)
+        future = future if future is None else torch.tensor(future, dtype=torch.float, device=device)
+
+        past_timestamp = torch.tensor(past_timestamp, dtype=torch.float, device=device)
+        future_timestamp = torch.tensor(future_timestamp, dtype=torch.float, device=device)
+
+        model_output = self.deep_model(past, past_timestamp, future, future_timestamp)
```

Comment:
After looking at it in more detail, I agree with Gerald. @yihaocs maybe you can simplify the rolling window dataset by removing the `start_token_len` parameter, and instead have each _model_ take `past[:, -start_token_len:]` if it needs to create a start token? Because in all cases, `future[:, :start_token_len]` should be equal to `past[:, -start_token_len]`. This also lets models choose their own behavior without relying on specific behavior from an external component (the dataset).